mpi3mr/storcli2.go

```diff
@@ -0,0 +1,631 @@
+	NumberOfBlocks               int    `json:"Number of Blocks,omitempty"`
+	SpanDepth                    int    `json:"Span Depth,omitempty"`
+	NumberOfDrives               int    `json:"Number of Drives,omitempty"`
+	DriveWriteCachePolicy        string `json:"Drive Write Cache Policy,omitempty"`
+	DefaultPowerSavePolicy       string `json:"Default Power Save Policy,omitempty"`
+	CurrentPowerSavePolicy       string `json:"Current Power Save Policy,omitempty"`
+	AccessPolicyStatus           string `json:"Access Policy Status,omitempty"`
+	AutoBGI                      string `json:"Auto BGI,omitempty"`
+	Secured                      string `json:"Secured,omitempty"`
+	InitState                    string `json:"Init State,omitempty"`
+	Consistent                   string `json:"Consistent,omitempty"`
+	Morphing                     string `json:"Morphing,omitempty"`
+	CachePreserved               string `json:"Cache Preserved,omitempty"`
+	BadBlockExists               string `json:"Bad Block Exists,omitempty"`
+	VDReadyForOSRequests         string `json:"VD Ready for OS Requests,omitempty"`
+	ReachedLDBBMFailureThreshold string `json:"Reached LD BBM failure threshold,omitempty"`
+	SupportedEraseTypes          string `json:"Supported Erase Types,omitempty"`
+	ExposedToOS                  string `json:"Exposed to OS,omitempty"`
+	CreationTimeLocalTimeString  string `json:"Creation Time(LocalTime yyyy/mm/dd hh:mm:sec),omitempty"`
+	DefaultCachebypassMode       string `json:"Default Cachebypass Mode,omitempty"`
+	CurrentCachebypassMode       string `json:"Current Cachebypass Mode,omitempty"`
+	SCSINAAId                    string `json:"SCSI NAA Id,omitempty"`
+	OSDriveName                  string `json:"OS Drive Name,omitempty"`
+	CurrentUnmapStatus           string `json:"Current Unmap Status,omitempty"`
+	CurrentWriteSameUnmapStatus  string `json:"Current WriteSame Unmap Status,omitempty"`
+	LUNSCountUsedPerPD           int    `json:"LU/NS Count used per PD,omitempty"`
+	DataFormatForIO              string `json:"Data Format for I/O,omitempty"`
+	SerialNumber                 string `json:"Serial Number,omitempty"`
+}
+
+func newController(cID int, cxShowNoLogJOut []byte, cxVAllShowAllJOut []byte) (Controller, error) {
+	ctrl := Controller{
+		ID: cID,
+	}
+
+	// parse Physical and Virtual Drives
+	s := StorCli2CmdCxShowResult{}
+	if err := json.Unmarshal(cxShowNoLogJOut, &s); err != nil {
+		return ctrl, err
+	}
+
+	// parse VirtDriveProperties
+	v := StorCli2CmdCxVallShowAllResult{}
+	if err := json.Unmarshal(cxVAllShowAllJOut, &v); err != nil {
+		return ctrl, err
+	}
+
+	// collect the physical drives
+	physDrives := []PhysicalDrive{}
+	foundPDrives := s.Controllers[cID].ResponseData.PhysicalDrives
+	physDrives = append(physDrives, foundPDrives...)
+
+	// collect the virtual drives
+	virtDrives := []VirtualDrive{}
+	foundVDrives := v.Controllers[cID].ResponseData.VirtualDrives // type VirtualDrives{}
```

Comment:
The difference between `VirtualDrives{}` and `VirtualDriveset{}` as a type that contains multiple virtual drives is confusing to me, at least on first read - I think it's that `VirtualDrives{}` is a type that's read from the input JSON directly and then virtualdriveset is constructed internally after munging?

Reply:
Yes.  The only difference between the Set and the Type, is that the set is mapping an "ID" to a VirtualDrives type.

I think the origin is that a "VirtualDrive" type is composed of physical drives, and could potentially be a mix of physical HDD and SSD, so the question for `VirtualDrive.IsSSD()` is decomposed into if all phys drives in a virt drive are SSD, then you can call the virt drive an SSD type.

But the mapping isn't super helpful now since the JSON output for a VirtualDrives type includes the PD info (each physical disk associated with a VirtualDrive) so we can easily enumerate the PDs from the structure.